File: go/recursion.go
package main

import "fmt"

func print(value int, counter *int) int {
	if value == 0 {
		return 0
	}
	var count int
	if counter == nil {
		count = 1
	} else {
		count = *counter + 1
	}
	fmt.Printf("%d ", count)
	return print(value-1, &count)
}

func main() {
	print(10, nil)
}

File: recursion.cpp
#include <iostream>
using namespace std;

void print(int value, int *counter) {
  if (value == 0) {
    return;
  }
  int count;
  if (counter == NULL) {
    count = 1;
  } else {
    count = *counter + 1;
  }
  std::cout << count << " ";
  return print(value - 1, &count);
}

int main() {
  print(10, NULL);
  return 0;
}
